docs/topics/addons.rst

```diff
@@ -0,0 +1,348 @@
+Afterwards, i.e. as soon as the :class:`~scrapy.addons.AddonManager` is
+populated, changing the ``ADDONS`` setting or any of the add-on
+configuration dictionary settings will have no effect.
+
+If you want to enable, disable, or configure add-ons in Python code, for example
+when writing your own add-on, you will have to use the
+:class:`~scrapy.addons.AddonManager`. You can access the add-on manager through
+either ``crawler.addons`` or, if you are writing an add-on, through the
+``addons`` argument of the :meth:`update_addons` callback. The add-on manager
+provides many useful methods and attributes to facilitate interacting with the
+add-ons framework, e.g.:
+
+* an :meth:`~scrapy.addons.AddonManager.add` method to load add-ons,
+* the :attr:`~scrapy.addons.AddonManager.enabled` list of enabled add-ons,
+* :meth:`~scrapy.addons.AddonManager.enable` and
+  :meth:`~scrapy.addons.AddonManager.disable` methods,
+* the :attr:`~scrapy.addons.AddonManager.configs` dictionary which holds the
+  configuration of all add-ons.
+
+In this example, we ensure that the ``httpcache`` add-on is loaded, and that
+its ``expiration_secs`` configuration is set to ``60``::
+
+    # addons is an instance of AddonManager
+    if 'httpcache' not in addons:
+        addons.add('httpcache', {'expiration_secs': 60})
+    else:
+        addons.configs['httpcache']['expiration_secs'] = 60
+
+
+Writing your own add-ons
+========================
+
+Add-ons are (any) Python *objects* that provide Scrapy's *add-on interface*.
+The interface is enforced through ``zope.interface``. This leaves the choice of
+Python object up the developer. Examples:
+
+* for a small pipeline, the add-on interface could be implemented in the same
+  class that also implements the ``open/close_spider`` and ``process_item``
+  callbacks,
+* for larger add-ons, or for clearer structure, the interface could be provided
+  by a stand-alone module.
+
+The absolute minimum interface consists of two attributes:
+
+.. attribute:: name
+
+    string with add-on name
+
+.. attribute:: version
+
+    version string (PEP-404, e.g. ``'1.0.1'``)
+
+Of course, stating just these two attributes will not get you very far. Add-ons
+can provide three callback methods that are called at various stages before the
+crawling process:
+
+.. method:: update_settings(config, settings)
+
+    This method is called during the initialization of the
+    :class:`~scrapy.crawler.Crawler`. Here, you should perform dependency checks
+    (e.g. for external Python libraries) and update the
+    :class:`~scrapy.settings.Settings` object as wished, e.g. enable components
+    for this add-on or set required configuration of other extensions.
+
+    :param config: Configuration of this add-on
+    :type config: ``dict``
+
+    :param settings: The settings object storing Scrapy/component configuration
+    :type settings: :class:`~scrapy.settings.Settings`
+
+.. method:: check_configuration(config, crawler)
```

Comment:
That's actually interesting; I wonder why does it accept crawler instead of settings, i.e. why does it check crawler state, not the final settings.

Reply:
I can see how having a crawler object available at the point in time when this method gets called could be useful, e.g. to monkeypatch some crawler component after it has been initialized.

That said, assuming addons are Scrapy components, meaning they can define a `from_crawler` class method and keep track of it (e.g. store it in  `self.crawler`), I think that and the use of signals should be able to cover most use cases there may be for having a crawler parameter in this method, or for having this method period :slightly_smiling_face:.